plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/portforwading/PortForwardingWizardModel.java

```diff
@@ -0,0 +1,190 @@
+import org.eclipse.ui.console.MessageConsole;
+import org.eclipse.ui.console.MessageConsoleStream;
+import org.jboss.tools.common.databinding.ObservablePojo;
+import org.jboss.tools.openshift.internal.common.ui.console.ConsoleUtils;
+
+import com.openshift.restclient.capability.CapabilityVisitor;
+import com.openshift.restclient.capability.resources.IPortForwardable;
+import com.openshift.restclient.model.IPod;
+import com.openshift.restclient.model.IPort;
+
+/**
+ * @author jeff.cantrill
+ */
+public class PortForwardingWizardModel extends ObservablePojo {
+
+	public static final String PROPERTY_FORWARDABLE_PORTS = "forwardablePorts";
+	public static final String PROPERTY_PORT_FORWARDING = "portForwarding";
+	public static final String PROPERTY_USE_FREE_PORTS = "useFreePorts";
+
+
+	private Boolean useFreePorts = Boolean.FALSE;
+	private final IPod pod;
+	private final ConsoleListener consoleListener = new ConsoleListener();
+	private List<IPortForwardable.PortPair> ports = new ArrayList<IPortForwardable.PortPair>();
+
+	public PortForwardingWizardModel(final IPod pod) {
+		this.pod = pod;
+		for (IPort port : pod.getContainerPorts()) {
+			ports.add(new IPortForwardable.PortPair(port));
+		}
+		
+	}
+
+	public final String getPodName() {
+		return pod.getName();
+	}
+
+	
+	public boolean getPortForwarding() {
+		return pod.accept(new CapabilityVisitor<IPortForwardable, Boolean>() {
+
+			@Override
+			public Boolean visit(IPortForwardable capability) {
+				return capability.isForwarding();
+			}
+		}, Boolean.FALSE);
+	}
+
+	public List<IPortForwardable.PortPair> getForwardablePorts(){
+		return Collections.unmodifiableList(ports);
+	}
+
+	public void startPortForwarding() {
+		pod.accept(new CapabilityVisitor<IPortForwardable, Object>() {
+
+			@Override
+			public Object visit(final IPortForwardable cap) {;
+				if(cap.isForwarding()) return null;
+				final MessageConsole console = ConsoleUtils.findMessageConsole(getMessageConsoleName());
+				consoleListener.setCapability(cap);
+				ConsoleUtils.registerConsoleListener(consoleListener);
+				MessageConsoleStream stream = console.newMessageStream();
+				stream.println("Starting port-forwarding...");
+				for (IPortForwardable.PortPair port : ports) {
+					stream.println(NLS.bind("{0} {1} -> {2}", new Object [] {port.getName(), port.getLocalPort(), port.getRemotePort()}));
+				}
+				cap.forwardPorts(ports.toArray(new IPortForwardable.PortPair []{}));
+				stream.println("done.");
+				ConsoleUtils.displayConsoleView(console);
+				firePropertyChange(PROPERTY_PORT_FORWARDING, false, getPortForwarding());
+				return null;
+			}
+		}, null);
+	}
+	
+	private class ConsoleListener implements IConsoleListener{
+		private IPortForwardable cap;
+		
+		public void setCapability(IPortForwardable cap) {
+			this.cap = cap;
+		}
+		@Override
+		public void consolesRemoved(IConsole[] consoles) {
+			final String messageConsoleName = getMessageConsoleName();
+			for (IConsole console : consoles) {
```

Comment:
This loop looks like an ideal candidate for a utils method in ConsoleUtils? ex. ConsoleUtils#containsNamedConsole(consoleName, IConsole consoles[])